walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private ArrayList<TimeRange> getMergedTimes(ArrayList<TimeRange> takenTimes) {
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    TimeRange previous = takenTimes.get(0);
+    ArrayList<TimeRange> takenTimesMerged = new ArrayList();
+
+    for (TimeRange time : takenTimes) {
+      if (time.overlaps(previous)) {
+        int start = previous.start();
+        int end = Math.max(time.end(), previous.end());
+        previous = TimeRange.fromStartEnd(start, end, false);
+      } else {
+          takenTimesMerged.add(previous);
+          previous = time;
+        }
+    }
+    takenTimesMerged.add(previous);
+    return takenTimesMerged;
+  }
+
+  private ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged) {
```

Comment:
What happens if "takenTimesMerged" is an empty list? I can see from your current implementation that that shouldn't be possible but if you had to make a change in the future you might not remember that this function will break if that undocumented expectation isn't met. Better to have this function robustly deal with an empty array.